[opt](rowset meta) truncate segments key bounds if too large to avoid `RowsetMetaCloudPB` exceeds fdb's 100KB limits

bobhan1 opened the pull request:
### What problem does this PR solve?

In cloud mode, `RowsetMetaCloudPB` will be stored as a value in fdb.  Segment key bounds is a member of `RowsetMetaCloudPB`. When the key columns has long data, segment key bounds may be very large, causing the size of `RowsetMetaCloudPB` be larger than the fdb 100KB limit, and will cause load txn fail when put `RowsetMetaCloudPB` KV in fdb.

This PR add a config `segments_key_bounds_truncation_threshold` to limit the max length of segment key bounds to avoid the problem. Segment key bounds which are larger than the value of this config will be truncated. The config will take effect when rowset is generated(load/schema change/compaction/snapshot).

**Note that as long as the config `segments_key_bounds_truncation_threshold` has ever been turned on, cluster downgrade and backup restore are not supported any more.**

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [x] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [x] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [x] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

zhannngchen commented on be/src/olap/base_tablet.h:
```diff
@@ -146,7 +146,8 @@ class BaseTablet {
                            OlapReaderStatistics& stats, std::string& values,
                            bool write_to_cache = false);
-
+    bool key_is_not_in_segment(Slice key, const KeyBoundsPB& segment_key_bounds,
```
the method name feels a bit unnatural; it should be called `is_key_in_segment`.
And it's better as a util method, not a member method of `BaseTablet`

zhannngchen replied:
Oh, I see. You can only confirm that the key is not in this segment, but you cannot confirm that it is in the segment.
You need to add comment for this method.

zhannngchen replied:
Oh, I see. You can only confirm that the key is not in this segment, but you cannot confirm that it is in the segment.
You need to add comment for this method.

zhannngchen commented on be/src/util/slice.cpp:
```diff
@@ -27,4 +27,23 @@ Slice::Slice(const faststring& s)
           size(s.size()) {}
 
+bool Slice::origin_is_strictly_less_than(Slice X, bool X_is_truncated, Slice Y,
+                                         bool Y_is_truncated) {
```
`Y_is_truncated` is useless?

gavinchou commented on gensrc/proto/olap_file.proto:
```diff
@@ -202,6 +203,8 @@ message RowsetMetaCloudPB {
     reserved 50;
     // to indicate whether the data between the segments overlap
     optional SegmentsOverlapPB segments_overlap_pb = 51 [default = OVERLAP_UNKNOWN];
+    optional int64 compaction_level = 52 [default = 0];
```
add comment for newly added fields

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run p0

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 committed:
upadte

remove reduntant config enable_segments_key_bounds_truncation

only set segments_key_bounds_truncated to true when actual truncate happend

fix

adjust

fix

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

dataroaring commented on be/src/olap/compaction.cpp:
```diff
@@ -105,11 +106,13 @@ bool is_rowset_tidy(std::string& pre_max_key, const RowsetSharedPtr& rhs) {
     }
-    if (min_key <= pre_max_key) {
+    bool cur_rs_key_bounds_truncated {rhs->is_segments_key_bounds_truncated()};
+    if (!Slice::origin_is_strictly_less_than(Slice {pre_max_key}, pre_rs_key_bounds_truncated,
```
```suggestion
    if (!Slice::lhs_is_strictly_less_than_rhs(Slice {pre_max_key}, pre_rs_key_bounds_truncated,
```

gavinchou commented on be/src/common/config.cpp:
```diff
@@ -1406,6 +1406,9 @@ DEFINE_mBool(enable_sleep_between_delete_cumu_compaction, "false");
 DEFINE_mInt32(compaction_num_per_round, "1");
 
+// the max length of segments key bounds, in bytes
+DEFINE_mInt32(segments_key_bounds_truncation_threshold, "100");
```
we should disable it by default => make it negative by default

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

zhannngchen reviewed (dismissed):
LGTM

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

gavinchou commented on be/src/common/config.cpp:
```diff
@@ -1493,6 +1493,9 @@ DEFINE_mBool(enable_compaction_pause_on_high_memory, "true");
 DEFINE_mBool(enable_calc_delete_bitmap_between_segments_concurrently, "false");
 
+// the max length of segments key bounds, in bytes
+DEFINE_mInt32(segments_key_bounds_truncation_threshold, "-1");
```
why not keep it a extremely large value to get rid of the check `segments_key_bounds_truncation_threshold > 0`

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `100.00% (73/73)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/be89b411923be362d981bad84406fc9205680416_be89b411923be362d981bad84406fc9205680416/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/be89b411923be362d981bad84406fc9205680416_be89b411923be362d981bad84406fc9205680416/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 54.81% (14777/26958) |
| Line Coverage     | 43.93% (129681/295180) |
| Region Coverage   | 42.63% (66182/155239) |
| Branch Coverage   | 37.23% (33402/89724) |

gavinchou commented on be/src/util/slice.h:
```diff
@@ -273,6 +273,13 @@ struct Slice {
+    // Y is (maybe) a truncated prefix of string Y'
+    // return true only if we can determine that X' is strictly less than Y'
+    // based on these maybe truncated prefixes
+    static bool lhs_is_strictly_less_than_rhs(Slice X, bool X_is_truncated, Slice Y,
```
the comment is confusing  and the parameters are not explained well
what is X and what is x_is_truncated.
it is better to show some examples to describe the behavior of this function.

gavinchou commented on be/src/util/key_util.cpp:
```diff
@@ -0,0 +1,31 @@
+
+namespace doris {
+
+bool key_is_not_in_segment(Slice key, const KeyBoundsPB& segment_key_bounds,
```
add UT for this function.
the UT is the explanation of the behavior of this function.

bobhan1 commented:
run external

bobhan1 commented:
run performance

bobhan1 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/e0f77d489371723b762ca06f65301c5d30ed0304_e0f77d489371723b762ca06f65301c5d30ed0304_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/e0f77d489371723b762ca06f65301c5d30ed0304_e0f77d489371723b762ca06f65301c5d30ed0304_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.15% (1105/1329) |
| Line Coverage     | 66.08% (18617/28175) |
| Region Coverage   | 65.65% (9268/14117) |
| Branch Coverage   | 55.31% (4979/9002) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `100.00% (73/73)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/45287_e0f77d489371723b762ca06f65301c5d30ed0304_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/45287_e0f77d489371723b762ca06f65301c5d30ed0304_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.45% (20870/26268) |
| Line Coverage     | 72.59% (214620/295679) |
| Region Coverage   | 70.75% (126214/178397) |
| Branch Coverage   | 64.46% (65360/101400) |

bobhan1 commented:
run buildall

bobhan1 committed:
add fuzzy conf

bobhan1 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/fdcd7c53356ee929f63dafa12a028742437ee35a_fdcd7c53356ee929f63dafa12a028742437ee35a_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/fdcd7c53356ee929f63dafa12a028742437ee35a_fdcd7c53356ee929f63dafa12a028742437ee35a_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.25% (1113/1337) |
| Line Coverage     | 66.27% (18701/28220) |
| Region Coverage   | 65.84% (9269/14077) |
| Branch Coverage   | 55.69% (4992/8964) |

bobhan1 commented:
run p0

bobhan1 commented:
run p0

bobhan1 commented:
run p0

bobhan1 commented:
run p0

bobhan1 committed:
log

bobhan1 committed:
fix log

bobhan1 committed:
update

bobhan1 committed:
fix case

bobhan1 committed:
fix

bobhan1 committed:
fix linked sc and add case

bobhan1 committed:
remove log

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/24af2a91113f6a0192c441456102d5628152c3b9_24af2a91113f6a0192c441456102d5628152c3b9_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/24af2a91113f6a0192c441456102d5628152c3b9_24af2a91113f6a0192c441456102d5628152c3b9_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.25% (1113/1337) |
| Line Coverage     | 66.15% (18670/28224) |
| Region Coverage   | 65.81% (9266/14079) |
| Branch Coverage   | 55.60% (4985/8966) |

bobhan1 committed:
add random_segments_key_bounds_truncation

bobhan1 commented:
run buildall

bobhan1 committed:
update

bobhan1 commented:
run buildall

bobhan1 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/12201b7d7b37157e5713fc19126d2e626ae4ba91_12201b7d7b37157e5713fc19126d2e626ae4ba91_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/12201b7d7b37157e5713fc19126d2e626ae4ba91_12201b7d7b37157e5713fc19126d2e626ae4ba91_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.26% (1114/1338) |
| Line Coverage     | 66.50% (18811/28287) |
| Region Coverage   | 66.14% (9324/14098) |
| Branch Coverage   | 56.08% (5034/8976) |

bobhan1 committed:
add comment

bobhan1 commented:
run buildall

bobhan1 commented:
run cloud_p0

dataroaring reviewed (approved):
LGTM

zhannngchen reviewed (approved):
LGTM
